app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java

```diff
@@ -546,4 +462,182 @@ public int setSelectionExpandWholeLines() {
     public boolean indexesValid(int... indexes) {
         return TextViewUtils.inRange(0, length(), indexes);
     }
+
+    static class LineNumbersDrawer {
+
+        private final AppCompatEditText _editor;
+        private final Paint _paint = new Paint();
+
+        private final int _defaultPaddingLeft;
+        private static final int LINE_NUMBER_PADDING_LEFT = 14;
+        private static final int LINE_NUMBER_PADDING_RIGHT = 10;
+
+        private final Rect _visibleArea = new Rect();
+        private final Rect _lineNumbersArea = new Rect();
+
+        private int _startNumber = 1;
+        private int _maxNumber = 1; // to gauge gutter width
+        private int _maxNumberDigits;
+        private int _lastMaxNumberDigits;
+        private int _lastLayoutLineCount;
+        private float _lastTextSize;
+        private int _numberX;
+        private int _gutterLineX;
+        private final List<Integer> _lineIndex = new ArrayList<>(); // Index of layout lines currently being drawn
+
+        // Current line numbers area
+        // Note: the distance between the top and bottom = 3 * height of visible area
+        // In order to reduce the frequency of refreshing parameters of line numbers,
+        // the line numbers are drawn on previous page, current page and next page every time
+        private int _top;
+        private int _bottom;
+
+
+        public LineNumbersDrawer(AppCompatEditText editor) {
+            _editor = editor;
+            _editor.getLocalVisibleRect(_lineNumbersArea);
+            _paint.setTextAlign(Paint.Align.RIGHT);
+            _defaultPaddingLeft = editor.getPaddingLeft();
+
+            _editor.addTextChangedListener(new GsTextWatcherAdapter() {
+
+                private final Pattern pattern = Pattern.compile("\n");
+                private Matcher matcher;
+
+                @Override
+                public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+                    if (after == 0 && count > 0) {
+                        CharSequence deleted = s.subSequence(start, start + count);
+                        matcher = pattern.matcher(deleted);
+                        while (matcher.find()) {
+                            _maxNumber--;
```

Comment:
Is tracking maxNumber really necessary? I don't in my version (I compute this on the fly in every onDraw) and there is not visible performance issue even on a very large file (10000 lines+)

Reply:
I think it is necessary. For example, a very large file (10000 lines+), when we view it at around 100th line, your method will still iterate from 1 to 10000+ (when executing the loop every time, and do this very frequently when scrolling), but my method will break the loop at around 100+, because it doesn't need to calculate the max line number.

Reply:
True., but I don't see a performance hit from this (I am running my branch on my phone for daily use). I can flick or drag scroll the entire text of Homer without any issues.

Remember that the textview has to layout the entire text anyway. Getting a line count is relatively trivial. 

I think optimizing for the performance at the end of the file (where one is more likely to view / edit) is what matters and both methods have the same performance here.